draft-ietf-quic-ack-frequency.md

```diff
@@ -335,17 +330,32 @@ send an acknowledgement immediately on receiving a reordered ack-eliciting
 packet. This extension modifies this behavior.
 
 If the endpoint has not yet received an ACK_FREQUENCY frame, or if the most
-recent frame received from the peer has an `Ignore Order` value of `false`
-(0x00), the endpoint MUST immediately acknowledge any subsequent packets that
+recent frame received from the peer has a `Reordering Threshold` value that is
+not 0x00, the endpoint MUST immediately acknowledge any subsequent packets that
 are received out of order.
 
-If the most recent ACK_FREQUENCY frame received from the peer has an `Ignore
-Order` value of `true` (0x01), the endpoint does not make this exception. That
+If the most recent ACK_FREQUENCY frame received from the peer has a `Reordering
+Threshold` value of 0x00, the endpoint does not make this exception. That
 is, the endpoint MUST NOT send an immediate acknowledgement in response to
 packets received out of order, and instead continues to use the peer's
 `Ack-Eliciting Threshold` and `max_ack_delay` thresholds for sending
 acknowledgements.
 
+### Reordering Threshold
+
+If the most recent value of 'Reordering Threshold' is not 0x00, then in addition
+to the immediate acknowledgement when reordering is detected, an additional
+immediate acknowledgement is sent when the peer can detect missing packets as
+lost.  This can speed up time threshold loss detection and therefore loss
+recovery.
+
+Every time an ACK frame is sent, the 'Largest Acknowledged' is saved in a local
+value (ie: largest_acknowledged_sent).  When a new packet is received in order,
+if there are any missing packets in the range
+'[largest_acknowledged_sent - Reordering Threshold,
+largest_acknowledged - Reordering Threshold],' send an immediate ACK.
```

Comment:
It took me several reads to understand what this was actually saying. I think an ", i.e. gap that are beyond the reordering threshold from latest received but not reported on in last ACK, then send an immediate ACK."

So If I understand this correctly, the receiver will re-iterate on gaps that are beyond threshold and thus considered lost as soon as they fall outside of the threshold. Thus a sender will not receive an immediate ack when things are actually just re-ordered and arrive to fill the gap as long as that happens within the threshold.

Reply:
I don't understand this as presented either.  Magnus' description fits better, maybe.

That would mean you are referring to the range from the Largest Acknowledged (i.e., the largest value in an ACK frame that has been sent) - Reordering Threshold to the Largest Received (i.e., the largest value in a packet that has been received) - Reordering Threshold.  That is [LA-RT, LR-RT].

I don't understand why you would look for gaps *before* the Largest Acknowledged here though.  Let's say the threshold is 2.  You receive packet 5 without getting packet 4.  That's a gap that would ordinarily elicit an immediate ACK.  But here you don't because it is newer than the 5-2=3 cutoff.  Then you wait for the time to pass and send an ACK.  Later, you receive packet 6, at which point your [LA-RT, LR-RT] range is [5-2=3, 6-2=4].  Presumably that includes packet 4[^1] so you send an immediate ACK.

That seems a bit pointless because you already informed the sender about that gap.  But maybe the sender logic won't detect loss until there are enough packets acknowledged beyond the gap, so this is useful.

You might instead want a range of [LA, LR-RT].  In that example, you would send an ACK on delay after packet 5 is received, meaning that the peer knows about the missing packet 4.  Later, when you receive 6, you don't send any more immediate ACKs.

The effect of either is that RT=1 has identical handling to RT=0.

[^1]: I don't know if you intentionally didn't say "inclusive", but I'm assuming that is why you are using brackets and not parentheses.  Please try to be explicit.

Reply:
I think this mixes two different mechanism here. The Ack-Eliciting Threshold may trigger an ack independent of the Reordering threshold. For the reordering threshold the only thing that is relevant is the number of packets received after the re-ordering/packet number gap was detected.

That means you have to have an internal flags that indicates that a gap was detected and count the number of packet received since. If during this time either the missing packet is received or an ACK is sent due to any other condition, the flag and counter get reset. If the counter becomes larger than the reordering threshold you send an Ack and also reset the flag and counter.